### ISSUE TITLE
rails3 link_to :method => :delete doesn't work correctly in IE8

### ISSUE THREAD

**Issue by inetufo** [rails/prototype-ujs#3]
<%= link_to 'Destroy', user, :confirm => 'Are you sure?', :method => :delete %>
When the confirm dialog shows, the request has already been posted and the destroy action is called, so the ok or cancel button have no effect, the record is always deleted,and i found out the request method is "post" not "delete".

this works fine in firefox

<%= csrf_meta_tag %> already in my layout file

**Comment by mislav** [comment]
The request method is "post" with a hidden parameter indicating it's a delete because raw "delete" HTTP requests are not widely supported.

As for the confirmation failure, I need to test it, but I think you are right that this is broken. Thanks for the report; I will post updates to this ticket

**Comment by inetufo** [comment]
Thanks for the reply, my environment is windows 7 64bit, ruby 1.8.7-p302 rails 3. I tested this on winxp in IE6, I clicked the "destroy" link, and the confirm dialog is shown,and at this time the request is not posted, so none of actions is called, but when i clicked the ok button, the "show" action is called not the "destroy" action as i expected.This really bothered me for a couple of hours, waiting for your answer.3x

**Comment by mislav** [comment]
Your server-side environment doesn't matter, since this is client-side code. All that matters is the browser you're using.

Is this new issue only present on IE6 and not on later browsers? Since this sounds like a separate issue, you should open a new ticket for it. Thanks

**Comment by inetufo** [comment]
I have tried this using jquery-ujs just now on win7 in IE8 and winxp in IE6, and everything goes well, i think it's a bug of prototype-ujs.

**Comment by inetufo** [comment]
Yeah,using prototype-ujs in IE8 it doesn't redirect to the "show" action,but the problem is the "destroy" action is called when the confirm dialog is shown, so the record is deleted before i click the ok or cancel button.Never tested in IE7, In firefox, this works fine.No other tests in any other browsers.

**Comment by ddeyoung** [comment]
I also have other remote links in IE 8 that are firing as a full request instead of a remote request using prototype-ujs.

**Comment by ddeyoung** [comment]
Replacing lines 138-153 of rails.js with the code below seems to make IE 8 behave:

```
  document.on("click", "*[data-confirm]", function(event, element) {
    if(element.readAttribute('data-remote') || element.readAttribute('data-method')) return;
    var message = element.readAttribute('data-confirm');
    if (!confirm(message)) event.stop();
  });

  document.on("click", "a[data-remote]", function(event, element) {
    if (event.stopped) return;
    if(element.readAttribute('data-confirm')) {
        var message = element.readAttribute('data-confirm');
        if (!confirm(message)) {
            event.stop();
            return;
        }
    }

    handleRemote(element);
    event.stop();
  });

  document.on("click", "a[data-method]", function(event, element) {
      if (event.stopped) return;
      if(element.readAttribute('data-remote')) return;
      if(element.readAttribute('data-confirm')) {
          var message = element.readAttribute('data-confirm');
          if (!confirm(message)) {
              event.stop();
              return;
          }
      }
      handleMethod(element);
      event.stop();
  });
```

**Comment by inetufo** [comment]
Thanks for your answer ddeyoung2.Replace with your code, the post request hasn't been sent when the confirm dialog shows.But new issue occus.The confirm dialog shows twice.Have you experinced this in your test with IE8?

I just commented the following codes 
//    document.on("click", "*[data-confirm]", function(event, element) {
//        if(element.readAttribute('data-remote') || element.readAttribute('date-method')) return;
//        var message = element.readAttribute('data-confirm');
//        if (!confirm(message)) event.stop();
//    });

and tried again in IE8, and everything goes well, i also tried it in firefox, and it behaves.

I don't know if commenting those codes above will generate new issue or not.

**Comment by inetufo** [comment]
Where is the author, Please have a look:)

**Comment by mislav** [comment]
@ddeyoung2: your code is a bit messy, but is the right direction I'll take for solving this. I'm going to take a look at this today/tomorrow, will close this ticket when solved. Thanks for all the info & help

**Comment by ddeyoung** [comment]
Whoops! There was a typo on line 2 'date-method' needs to be replaced with 'data-method'
We also cleaned up the code just a smidge ;) :

```
  document.on("click", "*[data-confirm]", function(event, element) {
    if(element.readAttribute('data-remote') || element.readAttribute('data-method')) return;
    var message = element.readAttribute('data-confirm');
    if (!confirm(message)) event.stop();
  });

  function confirmIfNeeded(element) {
      if(element.readAttribute('data-confirm')) {
            var message = element.readAttribute('data-confirm');
            if (!confirm(message)) {
                return false;
            }
        }
        return true; 
  }

  document.on("click", "a[data-remote]", function(event, element) {
    if (event.stopped) return;
    if(!confirmIfNeeded(element)) {
        event.stop();
        return;
    }

    handleRemote(element);
    event.stop();
  });

  document.on("click", "a[data-method]", function(event, element) {
      if (event.stopped) return;
      if(element.readAttribute('data-remote')) return;
      if(!confirmIfNeeded(element)) {
          event.stop();
          return;
      }
      handleMethod(element);
      event.stop();
  });
```

**Comment by inetufo** [comment]
Yeah, your new code works terrifically.I think the problem is solved.Thanks ddeyoung2.
Hope mislav update soon

**Comment by ddeyoung** [comment]
We did a little more consolidating: 

```
  document.on("click", "*[data-confirm]", function(event, element) {
    if(element.readAttribute('data-remote') || element.readAttribute('data-method')) return;
    var message = element.readAttribute('data-confirm');
    if (!confirm(message)) event.stop();
  });

  document.on("click", "a[data-remote], a[data-method]", function(event, element) {
    if (event.stopped) return;
    if(element.readAttribute('data-confirm')) {
        var message = element.readAttribute('data-confirm');
        if (!confirm(message)) {
            event.stop();
            return;
        }
    }
    if(element.readAttribute('data-remote')) {
        handleRemote(element);
    } else {
        handleMethod(element);
    }
    event.stop();
  });
```

**Comment by inetufo** [comment]
I have tried to solved this with jquery-ujs but new issues occurs.With jquery 1.4.1 this issue is ok, but with jquery 1.4.1 submit helper with :disable_with param doesn't work correctly, and i can't use rjs .Using jquery 1.4.3, it just redirects to the "show" action.I think it's a compatibility problem of IE8. Seems ok in firefox, no other browsers's tests like chrome and safari etc.Seems i have to come back to use prototype-ujs  :)

**Comment by mislav** [comment]
fix "data-confirm", "data-remote" not working reliably in IE

Refactored "data-confirm", "data-remote" and "data-method" handling on
forms and links not to rely on specific event handler ordering.

The issue manifested in IE as double confirmation dialogs on forms and
executing Ajax request regardless of confirmation outcome.

Closed by 60f352a18e68ec2ed97ae01b2d72cda9dda5a216
